madara/crates/client/block_import/src/verify_apply.rs

```diff
@@ -325,6 +328,52 @@ fn block_hash(
     Ok((block_hash, header))
 }
 
+/// Reorgs the blockchain from its current tip back to `new_tip` from which a new fork can then be
+/// played on top.
+///
+/// Returns the result of the reorg, which describes the part of the chain that was orphaned.
+#[allow(dead_code)] // not used yet outside of tests
+fn reorg(
+    backend: &MadaraBackend,
+    new_tip: &PreValidatedBlock, // TODO: we don't need a PreValidatedBlock, do we?
+) -> Result<ReorgResult, BlockImportError> {
+    let block_number = new_tip.unverified_block_number.expect("Can't reorg without block number");
+    // TODO: we should ensure this exact block exists in our db
+    let block_hash = new_tip.unverified_block_hash.expect("Can't reorg without a block hash");
+    let block_id = BasicId::new(block_number);
+
+    backend
+        .contract_trie()
+        .revert_to(block_id)
+        .map_err(|error| BlockImportError::Internal(Cow::Owned(format!("error reverting contract trie: {}", error))))?;
+
+    backend.contract_storage_trie().revert_to(block_id).map_err(|error| {
+        BlockImportError::Internal(Cow::Owned(format!("error reverting contract storage trie: {}", error)))
+    })?;
+
+    backend
+        .class_trie()
+        .revert_to(block_id)
+        .map_err(|error| BlockImportError::Internal(Cow::Owned(format!("error reverting class trie: {}", error))))?;
+
+    backend
+        .revert_to(block_number)
+        .map_err(|error| BlockImportError::Internal(Cow::Owned(format!("error reverting block db: {}", error))))?;
+
+    // TODO: ensure there is no race condition here...? (e.g. we reverted and someone else appends a block before this next call)
+    let latest_block_info = backend
+        .get_block_info(&BlockId::Tag(BlockTag::Latest))
+        .map_err(make_db_error("getting latest block info"))?
+        .ok_or(BlockImportError::Internal(Cow::Owned("no latest block after reorg".to_string())))?;
```

Comment:
Thoughts on this -- esp. whether it's an issue and how to prevent it?

Reply:
maybe using a lock for the entire reorg fn and any insert attempt should check that lock?
I think the scenario where a pending block is being inserted while a reorg happens is possible 🤔